rust/src/quic/frames.rs

```diff
@@ -0,0 +1,249 @@
+/* Copyright (C) 2020 Open Information Security Foundation
+ *
+ * You can copy, redistribute or modify this Program under the terms of
+ * the GNU General Public License version 2 as published by the Free
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+use super::error::QuicError;
+use nom::bytes::complete::take;
+use nom::number::complete::{be_u16, be_u8, le_u16, le_u32};
+use std::convert::TryFrom;
+use std::str::FromStr;
+use strum_macros::{Display, EnumString};
+
+/// Tuple of StreamTag and offset
+type TagOffset = (StreamTag, u32);
+
+/// Tuple of StreamTag and value
+type TagValue = (StreamTag, Vec<u8>);
+
+#[derive(Debug, PartialEq)]
+pub(crate) struct Stream {
+    pub fin: bool,
+    pub stream_id: Vec<u8>,
+    pub offset: Vec<u8>,
+    pub tags: Option<Vec<TagValue>>,
+}
+
+#[derive(Debug, PartialEq, Clone, Copy, Display, EnumString)]
+#[strum(serialize_all = "UPPERCASE")]
+pub(crate) enum StreamTag {
+    Aead,
+    Ccrt,
+    Ccs,
+    Cetv,
+    Cfcw,
+    Chlo,
+    Copt,
+    Csct,
+    Ctim,
+    Icsl,
+    Irtt,
+    Kexs,
+    Mids,
+    Mspc,
+    Nonc,
+    Nonp,
+    Pad,
+    Pdmd,
+    Pubs,
+    Scid,
+    Scls,
+    Sfcw,
+    Smhl,
+    Sni,
+    Sno,
+    Stk,
+    Tcid,
+    Uaid,
+    Ver,
+    Xlct,
+}
+
+#[derive(Debug, PartialEq)]
+pub(crate) enum Frame {
+    Padding,
+    Stream(Stream),
+    Unknown(Vec<u8>),
+}
+
+fn parse_tag(input: &[u8]) -> Result<(&[u8], StreamTag), QuicError> {
+    let (rest, tag) = take(4usize)(input)?;
+
+    let tag_as_str = std::str::from_utf8(tag)
+        .map_err(|e| {
+            QuicError::Parse(format!(
+                "Could not parse stream tag as str: {}",
+                e.to_string()
+            ))
+        })?
+        .trim_matches(char::from(0)); // trim null padding
+
+    let tag = StreamTag::from_str(tag_as_str).map_err(|e| {
```

Comment:
Maybe `http2_parse_frame_setting` syntax is easier with something like 
`map_opt!(be_u16, num::FromPrimitive::from_u16)` (u32 in your case)